tests/integrationv2/providers.py

```diff
@@ -308,8 +327,95 @@ def setup_server(self):
         return cmd_line
 
 
-class OpenSSL(Provider):
+class CriterionS2N(S2N):
+    """
+    Wrap the S2N provider in criterion-rust
+    """
+    criterion_off = 'off'
+    criterion_delta = 'delta'
+    criterion_baseline = 'baseline'
+    # Figure out what mode to run in: baseline, delta or report
+    criterion_mode = get_flag(S2N_USE_CRITERION)
+
+    def _find_s2n_benchmark(self, pattern):
+        # Use executable bit to find the correct file.
+        result = find_files(pattern, root_dir=self.cargo_root, mode='0o755')
+        if len(result) > 0:
+            return result[0]
+        else:
+            raise FileNotFoundError(
+                f"s2n criterion benchmark not found {result}.")
+
+    def _find_cargo(self):
+        # return a path to the highest level dir containing Cargo.toml
+        shortest = None
+        for file in find_files("Cargo.toml", root_dir="../.."):
+            if shortest is None:
+                shortest = file
+            else:
+                if len(file) < len(shortest):
+                    shortest = file
+        if shortest is None:
+            raise("Unable to find Cargo.toml")
+        # Return the path, minus Cargo.toml
+        return str("/".join(shortest.split("/")[:-1]))
+
+    def _cargo_bench(self):
+        """
+        Find the handlers
+        """
+        self.s2nc_bench = self._find_s2n_benchmark("s2nc-")
+        self.s2nd_bench = self._find_s2n_benchmark("s2nd-")
+
+    def __init__(self, options: ProviderOptions):
+        super().__init__(options)
+        # Set cargo root
+        self.cargo_root = self._find_cargo()
+
+        # Find the criterion binaries
+        self._cargo_bench()
+
+        # strip off the s2nc/d at the front because criterion
+        if 's2nc' in self.cmd_line[0] or 's2nd' in self.cmd_line[0]:
+            self.cmd_line = self.cmd_line[1:]
+
+        # strip off the s2nc -e argument, criterion handler isn't sending any STDIN characters,
+        # and makes it look like s2nc is hanging.
```

Comment:
How do the tests function like this? The "-e" means echo, and is how most of our tests either verify they succeeded or know to send more data.

Reply:
Talked to Doug offline. I was missing that he only updated the well known endpoints test, which doesn't send/receive any data. This is a blocker for pretty much every other test, but not that one.